Add more filters

n1k0 opened the pull request:
This patch adds filters for versions, langs and build targets.

![](http://i.imgur.com/ldkjcl2.png)

Deployed to [gh-pages](https://leplatrem.github.io/systemaddons-versions/) for easier QA.

n1k0 committed:
Factor filter checkbox as a reusable view.

n1k0 committed:
Add release lang filter.

n1k0 committed:
Add release target filters.

n1k0 committed:
Factor release prop extractors.

n1k0 committed:
Improve checkboxes styling.

n1k0 committed:
Add release version filters.

n1k0 commented:
I can't add @glasserc as a reviewer here but would appreciate benefiting from his valuable feedback :)

glasserc commented on ui/Model.elm:
```diff
@@ -126,42 +154,113 @@ getReleaseList =
+        extractFilterSet (.details >> .filename) releases
+            |> Dict.toList
+            |> List.map extractVersion
+            |> Dict.fromList
```
This is unfortunate. There's no `Dict.mapKeys` in the Elm core library, but maybe we should try to add one. (There is one in a dict-extra package, but I don't like it for reasons which are outside the scope of this discussion.)

n1k0 replied:
You mean contributing a `Dict.mapKeys` to Elm core? That's not something I'm planning to do anytime soon, mostly because I'm lazy. But feel free to work on this!

If you mean creating a local `mapDictKeys` function, I don't see the benefit as it's the only place where we'd need it.

glasserc commented on ui/Model.elm:
```diff
@@ -126,42 +154,113 @@ getReleaseList =
+
+
+processFilter : FilterSet -> (Release -> String) -> List Release -> List Release
+processFilter filterSet accessor releases =
```
This name is pretty opaque. It only makes sense in the context of `filterReleases` applying multiple `FilterSets`. Not sure what name would be better though.

glasserc commented on ui/Model.elm:
```diff
@@ -126,42 +154,113 @@ getReleaseList =
+        (\{ details } ->
+            List.any
+                (\f -> String.startsWith f details.filename)
+                (Dict.filter (\k v -> v) filterSet |> Dict.keys)
```
This is a bit clunky. Why not just re-use the logic to extract the version from `extractVersion` and use `processFilter`?

glasserc commented on ui/Model.elm:
```diff
@@ -126,42 +154,113 @@ getReleaseList =
+                newFilters =
+                    { filters
+                        | versions = Dict.update version (\_ -> Just active) filters.versions
                     }
```
These are all quite repetitive, why not extract them using a higher-order function?

n1k0 replied:
I couldn't find any way to update a record property dynamically in Elm, like `foo[key] = value` in JavaScript -- which is key for factoring these message handlers. I'll keep investigating though, and I can probably start with extracting a `toggleFilter` function for the dict update bit anyway.

glasserc replied:
Yeah, OK. In the Haskell community, there's a thing called "lenses" that people sometimes use for this sort of thing, but they are a bit intimidating for newbies and so they often feel like overkill. Elm doesn't really have any built-in way to do lenses so it's definitely not the right choice here. Maybe we could turn the filters into some kind of two-level structure, where the first level is "version", "lang", etc. and the second level is "50", "51", "52". But maybe that's also more trouble than it's worth. Anyhow I think the change you made here is good 👍

n1k0 committed:
@glasserc review, part 1: Factor filter toggling logic.

Moved filter update logic to its own update fn, created
dedicated toggling events.

n1k0 committed:
@glasserc review, part 2: Improve naming.

leplatrem commented:
Quick feedback:
* I would sort versions descending
* I would put channel filters on top
* The version is now shown in the details panel
* The title of each entry is currently the filename, I think we could just show `Firefox ${version}`

leplatrem commented:
Oh of course, another feedback: it's awesome!!

leplatrem commented on ui/Model.elm:
```diff
@@ -126,42 +159,100 @@ getReleaseList =
+                _ ->
+                    ( "None", x )
+    in
+        extractFilterSet (.details >> .filename) releases
```
Just wondering, why not use the version field instead?

n1k0 replied:
Fixed in https://github.com/leplatrem/systemaddons-versions/pull/2/commits/832a3bdede94cf68dc4975b60d4debcc0c0aebfa

n1k0 commented:
> I would sort versions descending

Funnily enough, `Dict`s are not sortable. Gonna have to use them as lists for ordering them.

n1k0 committed:
Extract the version from its dedicated prop.

n1k0 committed:
Order version by major DESC.

n1k0 committed:
Minor UI tweaks.

n1k0 committed:
Filters sidebar at a fixed position.

n1k0 committed:
Added addon id filter.

n1k0 commented:
I'm gonna land this. Build is broken because someone enabled CI for this repo while there's no .travis.yml just yet; we don't really care here.
